Modules/IO/ImageBase/test/itkWriteImageFunctionGTest.cxx

```diff
@@ -0,0 +1,71 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#include "itkImageFileWriter.h"
+#include "itkImage.h"
+
+#include "itkGTest.h"
+#include "itksys/SystemTools.hxx"
+#include "itkTestDriverIncludeRequiredIOFactories.h"
+
+#define STRING(s) #s
+namespace
+{
+
+class ITKWriteImageFunctionTest : public ::testing::Test
+{
+  void
+  SetUp() override
+  {
+    RegisterRequiredFactories();
+    itksys::SystemTools::ChangeDirectory(STRING(ITK_TEST_OUTPUT_DIR_STR));
+  }
+};
+
+using ImageType = itk::Image<float, 2>;
+typename ImageType::Pointer
+MakeImage()
+{
+  auto image = ImageType::New();
+
+  typename ImageType::RegionType region({ 100, 99 });
+
+  image->SetRegions(region);
+
+  image->Allocate();
+  return image;
+}
+
+} // namespace
+
+TEST_F(ITKWriteImageFunctionTest, ImageTypes)
+{
+  ImageType::Pointer image_ptr = MakeImage();
+  itk::WriteImage(image_ptr, "test1.mha");
+  itk::WriteImage(std::move(image_ptr), "test1.mha");
+
+  ImageType::ConstPointer image_cptr = MakeImage();
+  itk::WriteImage(image_cptr, "test1.mha");
+  itk::WriteImage(std::move(image_cptr), "test1.mha");
+
+  image_ptr = MakeImage();
```

Comment:
I don't think this reinitialization is necessary.

Reply:
Following best practices, after `std::move` on an object is should not be used again. It's state is unspecified.

Reply:
The `std::move` test also appears useful to me, to ensure that an r-value reference as first argument is properly supported. Especially because it might not have been supported if the parameter would have been declared as an l-value reference, `TImagePointer &` (with just a single ampersand).